### PR TITLE
Update the minimum external LLVM to 17

### PR DESCRIPTION
**Pull request by cuviper**
With this change, we'll have stable support for LLVM 17 and 18.
For reference, the previous increase to LLVM 16 was #117947.

### CODE REVIEW

**Comment by cuviper** [comment]
cc @rust-lang/wg-llvm  @RalfJung -- I believe this will unblock #117329.

And with my distro hat on, we needed LLVM 16 for Fedora 38, but that will be EOL before 1.79 is released.

**Comment by Kobzol** [comment]
I assume that all the LLVM version numbers in the description should be incremented by one? :)

**Comment by cuviper** [comment]
> I assume that all the LLVM version numbers in the description should be incremented by one? :)

D'oh! Fixed.

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile [hunk 1]
```diff
@@ -50,6 +53,10 @@ ENV RUST_CONFIGURE_ARGS \
       --enable-llvm-link-shared \
       --set rust.thin-lto-import-instr-limit=10
 
-COPY host-x86_64/x86_64-gnu-llvm-16/script.sh /tmp/
+COPY host-x86_64/dist-x86_64-linux/shared.sh /scripts/
+COPY host-x86_64/dist-x86_64-linux/build-gccjit.sh /scripts/
+
+RUN /scripts/build-gccjit.sh /scripts
```
Comment (nikic) [inline src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile:59]: Why these changes?

Comment (cuviper) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile:59]: That's syncing with changes that were made in 16 and 17, but not 18 yet -- I presume because they were developed around the same time 18 was added. But I think all of that is in question in #122546.

**Comment by Kobzol** [comment]
CI (and other) changes look good to me :+1:
